### DISCUSSION TITLE
how to use call dart from rust with my argument?

### DISCUSSION THREAD

**Discussion by Sherlock-Holo** [fzyzcjy/flutter_rust_bridge#1931] [Q&A] [answered]
I am reading the doc and I notice frb v2 supports call dart codes from rust side https://cjycode.com/flutter_rust_bridge/guides/direction/rust-call-dart

however it make me confuse

```dart
rustFunction(dartCallback: (name) => 'Hello, $name!');
```

that may mean set the callback in dart codes, so rust side can call it

but in rust side

```rust
pub fn rust_function(dart_callback: Fn(String) -> String) {
    dart_callback("Tom"); // Will get `Hello, Tom!`
}
```

well I don't know how to use this function

```rust
// other codes

rust_function(???where can I find the closure???)
```

or should I use something generated by frb?

and

if I want to pass my `name` variable to dart side, how to do it?

```rust
let name = "test".to_string();

// how to pass the variable as argument to function???
```

**Comment by fzyzcjy** [comment]
Hi, I am not very sure about the question. Could you please elaborate a bit more?

To be more explicitly, in case it helps:

```rust
pub fn rust_function(dart_callback: Fn(String) -> String) {
    let name = "test".to_string();
    dart_callback(name); // <-- answer your last question
}
```

```dart
void thisIsDartFunction(String name) { print('hi $name'); } // this is a dart function

rustFunction(dartCallback: thisIsDartFunction); // you can pass the dart function as an argument to Rust
```

**Reply by Sherlock-Holo** [reply, marked as answer]
oh I misunderstood the example

I thought

```dart
rustFunction(dartCallback: (name) => 'Hello, $name!');
```

meant dart provide a ffi function directly, like

```rust
fn my_rust_function(a: String) -> String { a.repeat(2) }
```

but actually: rust provide a function that allow dart call rust function to pass a callback to rust side, then rust can store the callback, and invoke the callback later to achieve the purpose of calling the dart codes
